src/search/parser/decorated_abstract_syntax_tree.cc

```diff
@@ -0,0 +1,563 @@
+#include "decorated_abstract_syntax_tree.h"
+
+#include "errors.h"
+
+#include "../plugins/construct_context.h"
+#include "../plugins/options.h"
+#include "../plugins/types.h"
+#include "../utils/logging.h"
+#include "../utils/math.h"
+#include "../utils/memory.h"
+
+#include <limits>
+
+using namespace std;
+
+namespace parser {
+void ConstructContext::push_layer(const string &layer) {
+    traceback.push(layer);
+}
+
+void ConstructContext::pop_layer() {
+    traceback.pop();
+}
+
+NO_RETURN
+void ConstructContext::construction_error(const string &message) const {
+    throw ParserError(message, traceback);
+}
+
+void ConstructContext::construction_warning(const string &message) const {
+    utils::g_log << traceback.str() << endl << message;
+}
+
+void ConstructContext::set_variable(const string &name, const plugins::Any &value) {
```

Comment:
Yes, the value is copied. We could move it here but there are other places (in `Options`) where we have to copy whole options objects, so I think we have to make the assumptions that types stored in objects are light-weight enough to be copied and passed by value. I left the code as it is because the move would add some complexity.